Init gradle extension plugin documentation (#20234)
quarkusio/quarkus: docs/src/main/asciidoc/writing-extensions.adoc

// The following link should point to the mojo page once https://github.com/quarkusio/quarkusio.github.io/issues/265 is fixed
Please refer to https://github.com/quarkusio/quarkus/blob/{quarkus-version}/devtools/maven/src/main/java/io/quarkus/maven/CreateExtensionMojo.java[CreateExtensionMojo JavaDoc] for all the available options of the mojo.

==== Using Gradle

> You will need to apply the `io.quarkus.extension` plugin in your `runtime` module and declare `quarkus-extension-processor` as annotation processor:

Reviewer: ```suggestion
You will need to apply the `io.quarkus.extension` plugin and enable annotation processor `io.quarkus:quarkus-extension-processor` in the `runtime` module of your extension project:
```

Reviewer: BTW, the annotation processor could be added by the plugin itself, right?

Reviewer: Are recommending to `enforcePlatform("io.quarkus:quarkus-bom:version")` somewhere?

Author: yes, for the runtime module I will push a commit to add it, however, for the deployment module, it could be tricky to find the module.

Author: I will add a `platform("io.quarkus:quarkus-bom:version")` reference in example. We can't use `enforcedPlatform` as it will block the publish.

Reviewer: Aren't you generating the `META-INF/quarkus-extension.properties` with the deployment artifact coordinates? Couldn't we find the deployment subproject and enable the annotation processor in it?

Author: Yes, and this is a property of the first task, maybe I could move it to an extension to have global plugin configuration available a configuration time.

> The plugin adds a `extensionDescriptor` task that will generate the `quarkus-extension.properties` file.

Reviewer: ```suggestion
The plugin includes the `extensionDescriptor` task that will generate the `META-INF/quarkus-extension.properties` file.
```

[source,groovy]
----
plugins {
    id 'java'
    id 'io.quarkus.extensions'
}

dependencies {
    annotationProcessor 'io.quarkus:quarkus-extension-processor'
}
----

> In your deployment module, you need to add `quarkus-extension-processor` as annotation processor:

Reviewer: ```suggestion
In your deployment module, you will need to enable only the annotation processor `io.quarkus:quarkus-extension-processor`:
```

[source,groovy]
----
plugins {
    id 'java'
}

dependencies {
    annotationProcessor 'io.quarkus:quarkus-extension-processor'
}
----

[WARNING]
====

> This plugin is still experimental. It will generate the `quarkus-extension.properties` file but not the `quarkus-extension.yml` file.

Reviewer: ```suggestion
This plugin is still experimental. It will generate the `META-INF/quarkus-extension.properties` file but not the `META-INF/quarkus-extension.yml` file.
```

> Also, it does not perform any check on your extension dependencies as the maven plugin do. Unit tests are not supported by the plugin for the moment.

Reviewer: ```suggestion
Also, it does not validate the extension dependencies as the equivalent Maven plugin does. Unit tests are not supported by the plugin for the moment.
```
